man/pylint.1
.TH pylint 1 "2020-03-13" pylint
.SH NAME
.B pylint
\- python code static checker

.SH SYNOPSIS
.B  pylint
[
.I OPTIONS
] [
.I <arguments>
]

.SH DESCRIPTION
.B pylint
is a Python source code analyzer which looks for programming
errors, helps enforcing a coding standard and sniffs for some code
smells (as defined in Martin Fowler's Refactoring book)

Pylint can be seen as another PyChecker since nearly all tests you
can do with PyChecker can also be done with Pylint. However, Pylint
offers some more features, like checking length of lines of code,
checking if variable names are well-formed according to your coding
standard, or checking if declared interfaces are truly implemented,
and much more.

Additionally, it is possible to write plugins to add your own checks.

Pylint is shipped with "pyreverse" (UML diagram generator)
and "symilar" (an independent similarities checker).

.SH OPTIONS
.IP "--version"
show program's version number and exit
.IP "--help, -h"
show this help message and exit
.IP "--long-help"
more verbose help.

.SH MASTER
.IP "--init-hook=<code>"
Python code to execute, usually for sys.path manipulation such as pygtk.require().
.IP "--errors-only, -E"
In error mode, checkers without error messages are disabled and for others, only the ERROR messages are displayed, and no reports are done by default.
.IP "--py3k"
In Python 3 porting mode, all checkers will be disabled and only messages emitted by the porting checker will be displayed.
.IP "--verbose, -v"
In verbose mode, extra non-checker-related info will be displayed.
.IP "--ignore=<file>[,<file>...]"
Files or directories to be skipped. They should be base names, not paths. [default: CVS]
.IP "--ignore-patterns=<pattern>[,<pattern>...]"
Files or directories matching the regex patterns are skipped. The regex matches against base names, not paths. [default: none]
.IP "--persistent=<y_or_n>"
Pickle collected data for later comparisons. [default: yes]
.IP "--load-plugins=<modules>"
List of plugins (as comma separated values of python module names) to load, usually to register additional checkers. [default: none]
.IP "--fail-under=<score>"
Specify a score threshold to be exceeded before program exits with error. [default: 10]
.IP "--jobs=<n-processes>, -j <n-processes>"
Use multiple processes to speed up Pylint. Specifying 0 will auto-detect the number of processors available to use. [default: 1]
.IP "--limit-inference-results=<number-of-results>"
Control the amount of potential inferred values when inferring a single object. This can help the performance when dealing with large functions or complex, nested conditions.  [default: 100]
.IP "--extension-pkg-allow-list=<pkg[,pkg]>"
A comma-separated list of package or module names from where C extensions may be loaded. Extensions are loading into the active Python interpreter and may run arbitrary code. [default: none]
.IP "--extension-pkg-whitelist=<pkg[,pkg]>"
A comma-separated list of package or module names from where C extensions may be loaded. Extensions are loading into the active Python interpreter and may run arbitrary code. (This is an alternative name to extension-pkg-allow-list for backward compatibility.) [default: none]
.IP "--suggestion-mode=<yn>"
When enabled, pylint would attempt to guess common misconfiguration and emit user-friendly hints instead of false-positive error messages. [default: yes]
.IP "--exit-zero"
Always return a 0 (non-error) status code, even if lint errors are found. This is primarily useful in continuous integration scripts.
.IP "--from-stdin"
Interpret the stdin as a python script, whose filename needs to be passed as the module_or_package argument.

.SH COMMANDS
.IP "--rcfile=<file>"
Specify a configuration file to load.
.IP "--help-msg=<msg-id>"
Display a help message for the given message id and exit. The value may be a comma separated list of message ids.
.IP "--list-msgs"
Generate pylint's messages.
.IP "--list-msgs-enabled"
Display a list of what messages are enabled and disabled with the given configuration.
.IP "--list-groups"
List pylint's message groups.
.IP "--list-conf-levels"
Generate pylint's confidence levels.
.IP "--full-documentation"
Generate pylint's full documentation.
.IP "--generate-rcfile"
Generate a sample configuration file according to the current configuration. You can put other options before this one to get them in the generated configuration.

.SH MESSAGES CONTROL
.IP "--confidence=<levels>"
Only show warnings with the listed confidence levels. Leave empty to show all. Valid levels: HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED. [default: none]
.IP "--enable=<msg ids>, -e <msg ids>"
Enable the message, report, category or checker with the given id(s). You can either give multiple identifier separated by comma (,) or put this option multiple time (only on the command line, not in the configuration file where it should appear only once). See also the "--disable" option for examples.
.IP "--disable=<msg ids>, -d <msg ids>"
Disable the message, report, category or checker with the given id(s). You can either give multiple identifiers separated by comma (,) or put this option multiple times (only on the command line, not in the configuration file where it should appear only once). You can also use "--disable=all" to disable everything first and then reenable specific checks. For example, if you want to run only the similarities checker, you can use "--disable=all --enable=similarities". If you want to run only the classes checker, but have no Warning level messages displayed, use "--disable=all --enable=classes --disable=W".

.SH REPORTS
.IP "--output-format=<format>, -f <format>"
Set the output format. Available formats are text, parseable, colorized, json and msvs (visual studio). You can also give a reporter class, e.g. mypackage.mymodule.MyReporterClass. [default: text]
.IP "--reports=<y_or_n>, -r <y_or_n>"
Tells whether to display a full report or only the messages. [default: no]
.IP "--evaluation=<python_expression>"
Python expression which should return a score less than or equal to 10. You have access to the variables 'error', 'warning', 'refactor', and 'convention' which contain the number of messages in each category, as well as 'statement' which is the total number of statements analyzed. This score is used by the global evaluation report (RP0004). [default: 10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)]
.IP "--score=<y_or_n>, -s <y_or_n>"
Activate the evaluation score. [default: yes]
.IP "--msg-template=<template>"
Template used to display messages. This is a python new-style format string used to format the message information. See doc for all details.

.SH LOGGING
.IP "--logging-modules=<comma separated list>"
Logging modules to check that the string format arguments are in logging function parameter format. [default: logging]
.IP "--logging-format-style=<old (%) or new ({) or fstr (f'')>"
Format style used to check logging format string. `old` means using % formatting, `new` is for `{}` formatting,and `fstr` is for f-strings. [default: old]

.SH SPELLING
.IP "--spelling-dict=<dict name>"
Spelling dictionary name. Available dictionaries: none. To make it work, install the python-enchant package. [default: none]
.IP "--spelling-ignore-words=<comma separated words>"
List of comma separated words that should not be checked. [default: none]
.IP "--spelling-private-dict-file=<path to file>"
A path to a file that contains the private dictionary; one word per line. [default: none]
.IP "--spelling-store-unknown-words=<y_or_n>"
Tells whether to store unknown words to the private dictionary (see the --spelling-private-dict-file option) instead of raising a message. [default: no]
.IP "--max-spelling-suggestions=N"
Limits count of emitted suggestions for spelling mistakes. [default: 4]

.SH MISCELLANEOUS
.IP "--notes=<comma separated values>"
List of note tags to take in consideration, separated by a comma. [default: FIXME,XXX,TODO]
.IP "--notes-rgx=<regexp>"
Regular expression of note tags to take in consideration.

.SH TYPECHECK
.IP "--ignore-on-opaque-inference=<y_or_n>"
This flag controls whether pylint should warn about no-member and similar checks whenever an opaque object is returned when inferring. The inference can return multiple potential results while evaluating a Python object, but some branches might not be evaluated, which results in partial inference. In that case, it might be useful to still emit no-member and other checks for the rest of the inferred objects. [default: yes]
.IP "--ignore-mixin-members=<y_or_n>"
Tells whether missing members accessed in mixin class should be ignored. A mixin class is detected if its name ends with "mixin" (case insensitive). [default: yes]
.IP "--ignore-none=<y_or_n>"
Tells whether to warn about missing members when the owner of the attribute is inferred to be None. [default: yes]
.IP "--ignored-modules=<module names>"
List of module names for which member attributes should not be checked (useful for modules/projects where namespaces are manipulated during runtime and thus existing member attributes cannot be deduced by static analysis). It supports qualified module names, as well as Unix pattern matching. [default: none]
.IP "--ignored-classes=<members names>"
List of class names for which member attributes should not be checked (useful for classes with dynamically set attributes). This supports the use of qualified names. [default: optparse.Values,thread._local,_thread._local]
.IP "--generated-members=<members names>"
List of members which are set dynamically and missed by pylint inference system, and so shouldn't trigger E1101 when accessed. Python regular expressions are accepted. [default: none]
.IP "--contextmanager-decorators=<decorator names>"
List of decorators that produce context managers, such as contextlib.contextmanager. Add to this list to register other decorators that produce valid context managers. [default: contextlib.contextmanager]
.IP "--missing-member-hint-distance=<member hint edit distance>"
The minimum edit distance a name should have in order to be considered a similar match for a missing member name. [default: 1]
.IP "--missing-member-max-choices=<member hint max choices>"
The total number of similar names that should be taken in consideration when showing a hint for a missing member. [default: 1]
.IP "--missing-member-hint=<missing member hint>"
Show a hint with possible names when a member name was not found. The aspect of finding the hint is based on edit distance. [default: yes]
.IP "--signature-mutators=<decorator names>"
List of decorators that change the signature of a decorated function. [default: none]

.SH VARIABLES
.IP "--init-import=<y_or_n>"
Tells whether we should check for unused import in __init__ files. [default: no]
.IP "--dummy-variables-rgx=<regexp>"
A regular expression matching the name of dummy variables (i.e. expected to not be used). [default: _+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_]
.IP "--additional-builtins=<comma separated list>"
List of additional names supposed to be defined in builtins. Remember that you should avoid defining new builtins when possible. [default: none]
.IP "--callbacks=<callbacks>"
List of strings which can identify a callback function by name. A callback name must start or end with one of those strings. [default: cb_,_cb]
.IP "--redefining-builtins-modules=<comma separated list>"
List of qualified module names which can have objects that can redefine builtins. [default: six.moves,past.builtins,future.builtins,builtins,io]
.IP "--ignored-argument-names=<regexp>"
Argument names that match this expression will be ignored. Default to name with leading underscore. [default: _.*|^ignored_|^unused_]
.IP "--allow-global-unused-variables=<y_or_n>"
Tells whether unused global variables should be treated as a violation. [default: yes]

.SH REFACTORING
.IP "--max-nested-blocks=<int>"
Maximum number of nested blocks for function / method body [default: 5]
.IP "--never-returning-functions=NEVER_RETURNING_FUNCTIONS"
Complete name of functions that never returns. When checking for inconsistent-return-statements if a never returning function is called then it will be considered as an explicit return statement and no message will be printed. [default: sys.exit]

.SH FORMAT
.IP "--max-line-length=<int>"
Maximum number of characters on a single line. [default: 100]
.IP "--ignore-long-lines=<regexp>"
Regexp for a line that is allowed to be longer than the limit. [default: ^\\s*(# )?<?https?://\\S+>?$]
.IP "--single-line-if-stmt=<y_or_n>"
Allow the body of an if to be on the same line as the test if there is no else. [default: no]
.IP "--single-line-class-stmt=<y_or_n>"
Allow the body of a class to be on the same line as the declaration if body contains single statement. [default: no]
.IP "--no-space-check=trailing-comma,dict-separator,empty-line"
List of optional constructs for which whitespace checking is disabled. `dict-separator` is used to allow tabulation in dicts, etc.: {1  : 1,\\n222: 2}. `trailing-comma` allows a space between comma and closing bracket: (a, ). `empty-line` allows space-only lines. [default: trailing-comma,dict-separator]
.IP "--max-module-lines=<int>"
Maximum number of lines in a module. [default: 1000]
.IP "--indent-string=<string>"
String used as indentation unit. This is usually "    " (4 spaces) or "\\t" (1 tab). [default: '    ']
.IP "--indent-after-paren=<int>"
Number of spaces of indent required inside a hanging or continued line. [default: 4]
.IP "--expected-line-ending-format=<empty or LF or CRLF>"
Expected format of line ending, e.g. empty (any line ending), LF or CRLF. [default: none]

.SH IMPORTS
.IP "--deprecated-modules=<modules>"
Deprecated modules which should not be used, separated by a comma. [default: optparse,tkinter.tix]
.IP "--preferred-modules=<module:preferred-module>"
Couples of modules and preferred modules, separated by a comma. [default: none]
.IP "--import-graph=<file.dot>"
Create a graph of every (i.e. internal and external) dependencies in the given file (report RP0402 must not be disabled). [default: none]
.IP "--ext-import-graph=<file.dot>"
Create a graph of external dependencies in the given file (report RP0402 must not be disabled). [default: none]
.IP "--int-import-graph=<file.dot>"
Create a graph of internal dependencies in the given file (report RP0402 must not be disabled). [default: none]
.IP "--known-standard-library=<modules>"
Force import order to recognize a module as part of the standard compatibility libraries.
.IP "--known-third-party=<modules>"
Force import order to recognize a module as part of a third party library. [default: enchant]
.IP "--allow-any-import-level=<modules>"
List of modules that can be imported at any level, not just the top level one. [default: none]
.IP "--analyse-fallback-blocks=<y_or_n>"
Analyse import fallback blocks. This can be used to support both Python 2 and 3 compatible code, which means that the block might have code that exists only in one or another interpreter, leading to false positives when analysed. [default: no]
.IP "--allow-wildcard-with-all=<y_or_n>"
Allow wildcard imports from modules that define __all__. [default: no]

.SH EXCEPTIONS
.IP "--overgeneral-exceptions=<comma-separated class names>"
Exceptions that will emit a warning when being caught. Defaults to "BaseException, Exception". [default: BaseException,Exception]

.SH CLASSES
.IP "--defining-attr-methods=<method names>"
List of method names used to declare (i.e. assign) instance attributes. [default: __init__,__new__,setUp,__post_init__]
.IP "--valid-classmethod-first-arg=<argument names>"
List of valid names for the first argument in a class method. [default: cls]
.IP "--valid-metaclass-classmethod-first-arg=<argument names>"
List of valid names for the first argument in a metaclass class method. [default: cls]
.IP "--exclude-protected=<protected access exclusions>"
List of member names, which should be excluded from the protected access warning. [default: _asdict,_fields,_replace,_source,_make]

.SH SIMILARITIES
.IP "--min-similarity-lines=<int>"
Minimum lines number of a similarity. [default: 4]
.IP "--ignore-comments=<y or n>"
Ignore comments when computing similarities. [default: yes]
.IP "--ignore-docstrings=<y or n>"
Ignore docstrings when computing similarities. [default: yes]
.IP "--ignore-imports=<y or n>"
Ignore imports when computing similarities. [default: no]

.SH DESIGN
.IP "--max-args=<int>"
Maximum number of arguments for function / method. [default: 5]
.IP "--max-locals=<int>"
Maximum number of locals for function / method body. [default: 15]
.IP "--max-returns=<int>"
Maximum number of return / yield for function / method body. [default: 6]
.IP "--max-branches=<int>"
Maximum number of branch for function / method body. [default: 12]
.IP "--max-statements=<int>"
Maximum number of statements in function / method body. [default: 50]
.IP "--max-parents=<num>"
Maximum number of parents for a class (see R0901). [default: 7]
.IP "--max-attributes=<num>"
Maximum number of attributes for a class (see R0902). [default: 7]
.IP "--min-public-methods=<num>"
Minimum number of public methods for a class (see R0903). [default: 2]
.IP "--max-public-methods=<num>"
Maximum number of public methods for a class (see R0904). [default: 20]
.IP "--max-bool-expr=<num>"
Maximum number of boolean expressions in an if statement (see R0916). [default: 5]

.SH BASIC
.IP "--good-names=<names>"
Good variable names which should always be accepted, separated by a comma. [default: i,j,k,ex,Run,_]
.IP "--good-names-rgxs=<names>"
Good variable names regexes, separated by a comma. If names match any regex, they will always be accepted [default: none]
.IP "--bad-names=<names>"
Bad variable names which should always be refused, separated by a comma. [default: foo,bar,baz,toto,tutu,tata]
.IP "--bad-names-rgxs=<names>"
Bad variable names regexes, separated by a comma. If names match any regex, they will always be refused [default: none]
.IP "--name-group=<name1:name2>"
Colon-delimited sets of names that determine each other's naming style when the name regexes allow several styles. [default: none]
.IP "--include-naming-hint=<y_or_n>"
Include a hint for the correct naming format with invalid-name. [default: no]
.IP "--property-classes=<decorator names>"
List of decorators that produce properties, such as abc.abstractproperty. Add to this list to register other decorators that produce valid properties. These decorators are taken in consideration only for invalid-name. [default: abc.abstractproperty]
.IP "--argument-naming-style=<style>"
Naming style matching correct argument names. [default: snake_case]
.IP "--argument-rgx=<regexp>"
Regular expression matching correct argument names. Overrides argument-naming-style.
.IP "--attr-naming-style=<style>"
Naming style matching correct attribute names. [default: snake_case]
.IP "--attr-rgx=<regexp>"
Regular expression matching correct attribute names. Overrides attr-naming-style.
.IP "--class-naming-style=<style>"
Naming style matching correct class names. [default: PascalCase]
.IP "--class-rgx=<regexp>"
Regular expression matching correct class names. Overrides class-naming-style.
.IP "--class-attribute-naming-style=<style>"
Naming style matching correct class attribute names. [default: any]
.IP "--class-attribute-rgx=<regexp>"
Regular expression matching correct class attribute names. Overrides class-attribute-naming-style.
.IP "--const-naming-style=<style>"
Naming style matching correct constant names. [default: UPPER_CASE]
.IP "--const-rgx=<regexp>"
Regular expression matching correct constant names. Overrides const-naming-style.
.IP "--function-naming-style=<style>"
Naming style matching correct function names. [default: snake_case]
.IP "--function-rgx=<regexp>"
Regular expression matching correct function names. Overrides function-naming-style.
.IP "--inlinevar-naming-style=<style>"
Naming style matching correct inline iteration names. [default: any]
.IP "--inlinevar-rgx=<regexp>"
Regular expression matching correct inline iteration names. Overrides inlinevar-naming-style.
.IP "--method-naming-style=<style>"
Naming style matching correct method names. [default: snake_case]
.IP "--method-rgx=<regexp>"
Regular expression matching correct method names. Overrides method-naming-style.
.IP "--module-naming-style=<style>"
Naming style matching correct module names. [default: snake_case]
.IP "--module-rgx=<regexp>"
Regular expression matching correct module names. Overrides module-naming-style.
.IP "--variable-naming-style=<style>"
Naming style matching correct variable names. [default: snake_case]
.IP "--variable-rgx=<regexp>"
Regular expression matching correct variable names. Overrides variable-naming-style.
.IP "--no-docstring-rgx=<regexp>"
Regular expression which should only match function or class names that do not require a docstring. [default: ^_]
.IP "--docstring-min-length=<int>"
Minimum line length for functions/classes that require docstrings, shorter ones are exempt. [default: -1]

.SH STRING
.IP "--check-str-concat-over-line-jumps=<y_or_n>"

ChangeLog
------------------
Pylint's ChangeLog
------------------

What's New in Pylint 2.8.0?
===========================
Release date: TBA

..
  Put new features here

* Introduce logic for checking deprecated attributes in DeprecationMixin.

* Reduce usage of blacklist/whitelist terminology. Notably, ``extension-pkg-allow-list`` is an
  alternative to ``extension-pkg-whitelist`` and the message ``blacklisted-name`` is now emitted as
  ``disallowed-name``. The previous names are accepted to maintain backward compatibility.


What's New in Pylint 2.7.3?
===========================
Release date: TBA

..
  Put bug fixes that will be cherry-picked to latest major version here

* Remove unwanted print to stdout from ``_emit_no_member``

* Introduce a command-line option to specify pyreverse output directory

  Closes #4159

* Fix issue with Enums and ``class-attribute-naming-style=snake_case``

  Closes #4149

* Add ``allowed-redefined-builtins`` option for fine tuning ``redefined-builtin`` check.

  Close #3263

* Fix issue when executing with ``python -m pylint``

  Closes #4161

* Exempt ``typing.TypedDict`` from ``too-few-public-methods`` check.

  Closes #4180

* Fix false-positive ``no-member`` for typed annotations without default value.

  Closes #3167

* Add ``--class-const-naming-style`` for Enum constants and class variables annotated
  with ``typing.ClassVar``

  Closes #4181

* Fix astroid.Inference error for undefined-variables with ``len()```

  Closes #4215


What's New in Pylint 2.7.2?
===========================
Release date: 2021-02-28

* Fix False Positive on ``Enum.__members__.items()``, ``Enum.__members__.values``, and ``Enum.__members__.keys``
  Closes #4123

* Properly strip dangerous sys.path entries (not just the first one)

  Closes #3636

* Workflow and packaging improvements


What's New in Pylint 2.7.1?
===========================
Release date: 2021-02-23

* Expose ``UnittestLinter`` in pylint.testutils

* Don't check directories starting with '.' when using register_plugins

  Closes #4119


What's New in Pylint 2.7.0?
===========================
Release date: 2021-02-21

* Introduce DeprecationMixin for reusable deprecation checks.

  Closes #4049

* Fix false positive for ``builtin-not-iterating`` when ``map`` receives iterable

  Closes #4078

* Python 3.6+ is now required.

* Fix false positive for ``builtin-not-iterating`` when ``zip`` receives iterable

* Add ``nan-comparison`` check for NaN comparisons

* Bug fix for empty-comment message line number.

  Closes #4009

* Only emit ``bad-reversed-sequence`` on dictionaries if below py3.8

  Closes #3940

* Handle class decorators applied to function.

  Closes #3882

* Add check for empty comments

* Fix minor documentation issue in contribute.rst

* Enums are now required to be named in UPPER_CASE by ``invalid-name``.

  Close #3834

* Add missing checks for deprecated functions.

* Postponed evaluation of annotations are now recognized by default if python version is above 3.10

  Closes #3992

* Fix column metadata for anomalous backslash lints

* Drop support for Python 3.5

* Add support for pep585 with postponed evaluation

  Closes #3320

* Check alternative union syntax - PEP 604

  Closes #4065

* Fix multiple false positives with assignment expressions

  Closes #3347, #3953, #3865, #3275

* Fix TypedDict inherit-non-class false-positive Python 3.9+

  Closes #1927

* Fix issue with nested PEP 585 syntax

* Fix issue with nested PEP 604 syntax

* Fix a crash in ``undefined-variable`` caused by chained attributes in metaclass

  Close #3742

* Fix false positive for ``not-async-context-manager`` when ``contextlib.asynccontextmanager`` is used

  Close #3862

* Fix linter multiprocessing pool shutdown (triggered warnings when runned in parallels with other pytest plugins)

  Closes #3779

* Fix a false-positive emission of ``no-self-use`` and ``unused-argument`` for methods
  of generic structural types (`Protocol[T]`)

  Closes #3885

* Fix bug that lead to duplicate messages when using ``--jobs 2`` or more.

  Close #3584

* Adds option ``check-protected-access-in-special-methods`` in the ClassChecker to activate/deactivate
  ``protected-access`` message emission for single underscore prefixed attribute in special methods.

  Close #3120

* Fix vulnerable regular expressions in ``pyreverse``

  Close #3811

* ``inconsistent-return-statements`` message is now emitted if one of ``try/except`` statement
  is not returning explicitly while the other do.

  Closes #3468

* Fix ``useless-super-delegation`` false positive when default keyword argument is a dictionnary.

  Close #3773

* Fix a crash when a specified config file does not exist

* Add support to ``ignored-argument-names`` in DocstringParameterChecker and adds ``useless-param-doc`` and ``useless-type-doc`` messages.

  Close #3800

* Enforce docparams consistently when docstring is not present

  Close #2738

* Fix ``duplicate-code`` false positive when lines only contain whitespace and non-alphanumeric characters (e.g. parentheses, bracket, comman, etc.)

* Improve lint message for ``singleton-comparison`` with bools

* Fix spell-checker crash on indented docstring lines that look like # comments

  Close #3786

* Fix AttributeError in checkers/refactoring.py

* Improve sphinx directives spelling filter

* Fix a bug with postponed evaluation when using aliases for annotations.

  Close #3798

* Fix minor documentation issues

* Improve the performance of the line length check.

* Removed incorrect deprecation of ``inspect.getfullargspec``

* Fix ``signature-differs`` false positive for functions with variadics

  Close #3737

* Fix a crash in ``consider-using-enumerate`` when encountering ``range()`` without arguments

  Close #3735


elisp/pylint.el
;;; pylint.el --- minor mode for running `pylint'

;; Copyright (c) 2009, 2010 Ian Eure <ian.eure@gmail.com>
;; Author: Ian Eure <ian.eure@gmail.com>
;; Maintainer: Jonathan Kotta <jpkotta@gmail.com>

;; Keywords: languages python
;; Version: 1.02

;; pylint.el is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by the Free
;; Software Foundation; either version 2, or (at your option) any later
;; version.
;;
;; It is distributed in the hope that it will be useful, but WITHOUT ANY
;; WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
;; FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
;; details.
;;
;; You should have received a copy of the GNU General Public License along
;; with your copy of Emacs; see the file COPYING.  If not, write to the Free
;; Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
;; MA 02110-1301, USA

;;; Commentary:
;;
;; Specialized compile mode for pylint.  You may want to add the
;; following to your init.el:
;;
;;   (autoload 'pylint "pylint")
;;   (add-hook 'python-mode-hook 'pylint-add-menu-items)
;;   (add-hook 'python-mode-hook 'pylint-add-key-bindings)
;;
;; There is also a handy command `pylint-insert-ignore-comment' that
;; makes it easy to insert comments of the form `# pylint:
;; ignore=msg1,msg2,...'.

;;; Code:

(require 'compile)
(require 'tramp)

(defgroup pylint nil
  "Minor mode for running the Pylint Python checker"
  :prefix "pylint-"
  :group 'tools
  :group 'languages)

(defvar pylint-last-buffer nil
  "The most recent PYLINT buffer.
A PYLINT buffer becomes most recent when you select PYLINT mode in it.
Notice that using \\[next-error] or \\[compile-goto-error] modifies
`completion-last-buffer' rather than `pylint-last-buffer'.")

(defconst pylint-regexp-alist
  (let ((base "^\\(.*\\):\\([0-9]+\\):\s+\\(\\[%s.*\\)$"))
    (list
     (list (format base "[FE]") 1 2)
     (list (format base "[RWC]") 1 2 nil 1)))
  "Regexp used to match PYLINT hits.  See `compilation-error-regexp-alist'.")

(defcustom pylint-options '("--reports=n" "--output-format=parseable")
  "Options to pass to pylint.py"
  :type '(repeat string)
  :group 'pylint)

(defcustom pylint-use-python-indent-offset nil
  "If non-nil, use `python-indent-offset' to set indent-string."
  :type 'boolean
  :group 'pylint)

(defcustom pylint-command "pylint"
  "PYLINT command."
  :type '(file)
  :group 'pylint)

(defcustom pylint-alternate-pylint-command "pylint2"
  "Command for pylint when invoked with C-u."
  :type '(file)
  :group 'pylint)

(defcustom pylint-ask-about-save nil
  "Non-nil means \\[pylint] asks which buffers to save before compiling.
Otherwise, it saves all modified buffers without asking."
  :type 'boolean
  :group 'pylint)

(defvar pylint--messages-list ()
  "A list of strings of all pylint messages.")

(defvar pylint--messages-list-hist ()
  "Completion history for `pylint--messages-list'.")

(defun pylint--sort-messages (a b)
  "Compare function for sorting `pylint--messages-list'.

Sorts most recently used elements first using `pylint--messages-list-hist'."
  (let ((idx 0)
        (a-idx most-positive-fixnum)
        (b-idx most-positive-fixnum))
    (dolist (e pylint--messages-list-hist)
      (when (string= e a)
        (setq a-idx idx))
      (when (string= e b)
        (setq b-idx idx))
      (setq idx (1+ idx)))
    (< a-idx b-idx)))

(defun pylint--create-messages-list ()
  "Use `pylint-command' to populate `pylint--messages-list'."
  ;; example output:
  ;;  |--we want this--|
  ;;  v                v
  ;; :using-cmp-argument (W1640): *Using the cmp argument for list.sort / sorted*
  ;;   Using the cmp argument for list.sort or the sorted builtin should be avoided,
  ;;   since it was removed in Python 3. Using either `key` or `functools.cmp_to_key`
  ;;   should be preferred. This message can't be emitted when using Python >= 3.0.
  (setq pylint--messages-list
        (split-string
         (with-temp-buffer
           (shell-command (concat pylint-command " --list-msgs") (current-buffer))
           (flush-lines "^[^:]")
           (goto-char (point-min))
           (while (not (eobp))
             (delete-char 1) ;; delete ";"
             (re-search-forward " ")
             (delete-region (point) (line-end-position))
             (forward-line 1))
           (buffer-substring-no-properties (point-min) (point-max))))))

;;;###autoload
(defun pylint-insert-ignore-comment (&optional arg)
  "Insert a comment like \"# pylint: disable=msg1,msg2,...\".

This command repeatedly uses `completing-read' to match known
messages, and ultimately inserts a comma-separated list of all
the selected messages.

With prefix argument, only insert a comma-separated list (for
appending to an existing list)."
  (interactive "*P")
  (unless pylint--messages-list
    (pylint--create-messages-list))
  (setq pylint--messages-list
        (sort pylint--messages-list #'pylint--sort-messages))
  (let ((msgs ())
        (msg "")
        (prefix (if arg
                    ","
                  "# pylint: disable="))
        (sentinel "[DONE]"))
    (while (progn
             (setq msg (completing-read
                        "Message: "
                        pylint--messages-list
                        nil t nil 'pylint--messages-list-hist sentinel))
             (unless (string= sentinel msg)
               (add-to-list 'msgs msg 'append))))
    (setq pylint--messages-list-hist
          (delete sentinel pylint--messages-list-hist))
    (insert prefix (mapconcat 'identity msgs ","))))

(define-compilation-mode pylint-mode "PYLINT"
  (setq pylint-last-buffer (current-buffer))
  (set (make-local-variable 'compilation-error-regexp-alist)
       pylint-regexp-alist)
  (set (make-local-variable 'compilation-disable-input) t))

(defvar pylint-mode-map
  (let ((map (make-sparse-keymap)))
    (set-keymap-parent map compilation-minor-mode-map)
    (define-key map " " 'scroll-up)
    (define-key map "\^?" 'scroll-down)
    (define-key map "\C-c\C-f" 'next-error-follow-minor-mode)

    (define-key map "\r" 'compile-goto-error)  ;; ?
    (define-key map "n" 'next-error-no-select)
    (define-key map "p" 'previous-error-no-select)
    (define-key map "{" 'compilation-previous-file)
    (define-key map "}" 'compilation-next-file)
    (define-key map "\t" 'compilation-next-error)
    (define-key map [backtab] 'compilation-previous-error)
    map)
  "Keymap for PYLINT buffers.
`compilation-minor-mode-map' is a cdr of this.")

(defun pylint--make-indent-string ()
  "Make a string for the `--indent-string' option."
  (format "--indent-string='%s'"
          (make-string python-indent-offset ?\ )))

;;;###autoload
(defun pylint (&optional arg)
  "Run PYLINT, and collect output in a buffer, much like `compile'.

While pylint runs asynchronously, you can use \\[next-error] (M-x next-error),
or \\<pylint-mode-map>\\[compile-goto-error] in the grep \
output buffer, to go to the lines where pylint found matches.

\\{pylint-mode-map}"
  (interactive "P")

  (save-some-buffers (not pylint-ask-about-save) nil)
  (let* ((filename (buffer-file-name))
         (filename (or (and (tramp-tramp-file-p filename)
                         (aref (tramp-dissect-file-name filename) 3))
                      filename))
         (filename (shell-quote-argument filename))
         (pylint-command (if arg
                             pylint-alternate-pylint-command
                           pylint-command))
         (pylint-options (if (not pylint-use-python-indent-offset)
                             pylint-options
                           (append pylint-options
                                   (list (pylint--make-indent-string)))))
         (command (mapconcat
                   'identity
                   (append `(,pylint-command) pylint-options `(,filename))
                   " ")))

    (compilation-start command 'pylint-mode)))

;;;###autoload
(defun pylint-add-key-bindings ()
  (let ((map (cond
              ((boundp 'py-mode-map) py-mode-map)
              ((boundp 'python-mode-map) python-mode-map))))

    ;; shortcuts in the tradition of python-mode and ropemacs
    (define-key map (kbd "C-c m l") 'pylint)
    (define-key map (kbd "C-c m p") 'previous-error)
    (define-key map (kbd "C-c m n") 'next-error)
    (define-key map (kbd "C-c m i") 'pylint-insert-ignore-comment)
    nil))

;;;###autoload
(defun pylint-add-menu-items ()
  (let ((map (cond
              ((boundp 'py-mode-map) py-mode-map)
              ((boundp 'python-mode-map) python-mode-map))))

    (define-key map [menu-bar Python pylint-separator]
      '("--" . pylint-separator))
    (define-key map [menu-bar Python next-error]
      '("Next error" . next-error))
    (define-key map [menu-bar Python prev-error]
      '("Previous error" . previous-error))
    (define-key map [menu-bar Python lint]
      '("Pylint" . pylint))
    nil))

(provide 'pylint)

;;; pylint.el ends here
